File: fabric-samples/chaincode/NewsCoin/go/NewsCoin.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * The sample smart contract for documentation topic:
 * Writing Your First Blockchain Application
 */

package main

/* Imports
 * 4 utility libraries for formatting, handling bytes, reading and writing JSON, and string manipulation
 * 2 specific Hyperledger Fabric specific libraries for Smart Contracts
 */
import (
	"encoding/json"
	"fmt"
	"strconv"
	"crypto"
	"crypto/rsa"
	"os"
	"encoding/hex"
	"crypto/sha256"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	sc "github.com/hyperledger/fabric/protos/peer"
)

// Define the Smart Contract structure
type SmartContract struct {
}

// Define the car structure, with 4 properties.  Structure tags are used by encoding/json library
type Peer struct {
	Value  			float64 	 `json:"value"`
	Public_key	string     `json:"public_key"`
}

/*
 * The Init method is called when the Smart Contract "fabcar" is instantiated by the blockchain network
 * Best practice is to have any Ledger initialization in separate function -- see initLedger()
 */
func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
	return shim.Success(nil)
}

/*
 * The Invoke method is called as a result of an application request to run the Smart Contract "fabcar"
 * The calling application program has also specified the particular smart contract function to be called, with arguments
 */
func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {

	// Retrieve the requested Smart Contract function and arguments
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "queryPeer" {
		return s.queryPeer(APIstub, args)
	} else if function == "initLedger" {
		return s.initLedger(APIstub)
	} else if function == "createPeer" {
		return s.createPeer(APIstub, args)
	} else if function == "transaction" {
		return s.transaction(APIstub, args)
	}

	return shim.Error("Invalid Smart Contract function name.")
}

func (s *SmartContract) queryPeer(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	peerAsBytes, _ := APIstub.GetState(args[0])
	return shim.Success(peerAsBytes)
}

func (s *SmartContract) initLedger(APIstub shim.ChaincodeStubInterface) sc.Response {
	peers := []Peer{
		Peer{Value: 500000000},
		Peer{Value: 500000000},
	}

	peerAsBytes1, _ := json.Marshal(peers[0])
	APIstub.PutState("MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCcmtaOSBALW50YeLBL4JdQUDyIyFnXoHnM2SaOrREVJVvSERthaCdZKRkpGNDB5yBdENf+NI9KCSkBdVK5Zj0QR8GDMxGwwPZcnLDeT+tNxpSnPRneJ8vvpiXO3FlWsPOeVhg6Mtz+qernMs/MNC5eV8TImnu9i4Yg0TMFqBGWQQIDAQAB", peerAsBytes1)
	fmt.Println("Added", peers[0])
	peerAsBytes2, _ := json.Marshal(peers[1])
	APIstub.PutState("MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCAIWNeitXAYWPeGHH7nxF0Ag5RBy7qdibpD2vjDMik1sZPFiRHdvpcl+Bm5OWSA1QxZufQ1cPgefhy9JUeujfmNXsKDEFJJVBJ0+L2Q8nSnv0fmXvtPnYfoMkjIWITNiMLHCqNyIdcFA7VR9DAjE87DbCDw+IPtFU4nhqabPxd5wIDAQAB", peerAsBytes2)
	fmt.Println("Added", peers[1])


	return shim.Success(nil)
}

func (s *SmartContract) createPeer(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	var peer = Peer{Value: 0}

	peerAsBytes, _ := json.Marshal(peer)
	APIstub.PutState(args[0], peerAsBytes)

	return shim.Success(nil)
}

func (s *SmartContract) transaction(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

		if len(args) != 4 {
			return shim.Error("Incorrect number of arguments. Expecting 4")
		}

		peerAsBytes1, _ := APIstub.GetState(args[0])
		peer1 := Peer{}

		peerAsBytes2, _ := APIstub.GetState(args[1])
		peer2 := Peer{}

		valor , _ := strconv.ParseFloat(args[2],64)

		json.Unmarshal(peerAsBytes1, &peer1)
		json.Unmarshal(peerAsBytes2, &peer2)

		message := []byte(peer1.Public_key)

		signature, _ := hex.DecodeString(args[3])

		hashed := sha256.Sum256(message)

		err := rsa.VerifyPKCS1v15(peer1.Public_key, crypto.SHA256, hashed[:], signature)
		if err != nil {
        fmt.Fprintf(os.Stderr, "Error from verification: %s\n", err)
		}

		if valor < 0.0000000001{
			return shim.Error("Quantia não suportada Min:0.0000000001")
		} else if peer1.Value > valor{
			peer1.Value -= valor
		} else {
		  return shim.Error("Saldo insuficiente")
		}

		peer2.Value += valor

		peerAsBytes1, _ = json.Marshal(peer1)
		APIstub.PutState(args[0], peerAsBytes1)

		peerAsBytes2, _ = json.Marshal(peer2)
		APIstub.PutState(args[1], peerAsBytes2)

		return shim.Success(nil)
}

// The main function is only relevant in unit test mode. Only included here for completeness.
func main() {

	// Create a new Smart Contract
	err := shim.Start(new(SmartContract))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}

File: fabric-samples/chaincode/NewsCoin/node/NewsCoin.js
/*
# Copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
*/

'use strict';
const shim = require('fabric-shim');
const util = require('util');
const crypto = require('crypto');

let Chaincode = class {

  // The Init method is called when the Smart Contract 'fabcar' is instantiated by the blockchain network
  // Best practice is to have any Ledger initialization in separate function -- see initLedger()
  async Init(stub) {
    console.info('=========== Instantiated fabcar chaincode ===========');
    return shim.success();
  }

  // The Invoke method is called as a result of an application request to run the Smart Contract
  // 'fabcar'. The calling application program has also specified the particular smart contract
  // function to be called, with arguments
  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);

    let method = this[ret.fcn];
    if (!method) {
      console.error('no function of name:' + ret.fcn + ' found');
      throw new Error('Received unknown function ' + ret.fcn + ' invocation');
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  async queryPeer(stub, args) {
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments');
    }
    let peer = args[0];

    let peerAsBytes = await stub.getState(peer);
    if (!peerAsBytes || peerAsBytes.toString().length <= 0) {
      throw new Error(peer + ' does not exist: ');
    }
    console.log(peerAsBytes.toString());
    let p = JSON.parse(peerAsBytes)
    return p.value.toString();
  }

  async initLedger(stub, args) {
    console.info('============= START : Initialize Ledger ===========');
    let peers = [];
    peers.push({
      value: '500000000',
      public_key: '-----BEGIN RSA PUBLIC KEY-----\n'+
      'MIIBCgKCAQEAhF6963IwAP8g2E7MC5Fr9yMQVkiuTFf5VT16pW4fhYcJtLEWw7N8BuS7GKVR\n'+
      '84vtJUsbNlgtbCpVnk8adS4DavtQvUdzBcBzIhpduUlP4iFQyIgpQN8yNAcXCiYbd0alTVxL\n'+
      '4n8TU7LixWkzPyfPdhuCcMCvSEU7v1caDdYyYmZwz8Nk19fptPyn90t6FH6aHx2MYH0ARTM5\n'+
      'cTs3kc0huUEDukPRjho97lHf2n5F2SChW7GdC21fH2Ix/Fj2UlAe8HgzD2iYnwkDfKVPUhGb\n'+
      'wK4uj9K+IA/6Ftnom8oAexuJAWwyktqUyaoSqTrZFSM4+26mTLbcTveVB8UVCvOrkwIDAQAB\n'+
      '-----END RSA PUBLIC KEY-----'
    });
    peers.push({
      value: '500000000',
      public_key: '-----BEGIN RSA PUBLIC KEY-----\n'+
      'MIIBCgKCAQEApq/4WP31aFm3kgl2CgIJwDugpi3XgLHstPdVnAzPKIZuSJWinUdaXRo4i8K7\n'+
      'hHW4VJf/wrhRqv57gg0E6nbnD/RFuw/8cdjrjj2+NOuA7Iz8k3Dwq1bR7/IqDIM/Hxp8RzT9\n'+
      '2X57XMxxCpIU8K5EEczc+nDxY2zuVGbCXlqtH+6/VIGDHovlHveVKmyT9wAn4E/Uv2HVOxcK\n'+
      'TnKGHtpcN3kjkzX0vs6NAGJqNixpp5fcQ8vxDUtlMLr2Kco+g6hzq72sRQIdgxOkpNt8B51Z\n'+
      '1/JbtKkaMyxuZRdwzHk/0flFAk/RmdJFloiyEnhCj0wp5pSGe3G36syCxCPfmD5FGQIDAQAB\n'+
      '-----END RSA PUBLIC KEY-----'
    });

      await stub.putState('MIIBCgKCAQEAhF6963IwAP8g2E7MC5Fr9yMQVkiuTFf5VT16pW4fhYcJtLEWw7N8BuS7GKVR'+
      '84vtJUsbNlgtbCpVnk8adS4DavtQvUdzBcBzIhpduUlP4iFQyIgpQN8yNAcXCiYbd0alTVxL'
      , Buffer.from(JSON.stringify(peers[0])));

      await stub.putState('MIIBCgKCAQEApq/4WP31aFm3kgl2CgIJwDugpi3XgLHstPdVnAzPKIZuSJWinUdaXRo4i8K7'+
      'hHW4VJf/wrhRqv57gg0E6nbnD/RFuw/8cdjrjj2+NOuA7Iz8k3Dwq1bR7/IqDIM/Hxp8RzT9'
      , Buffer.from(JSON.stringify(peers[0])));

    console.info('============= END : Initialize Ledger ===========');
  }

  async createPeer(stub, args) {
    console.info('============= START : Create Peer ===========');
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting 1');
    }

    var peer = {
      docType: 'peer',
      value: 0,
      public_key: args[0],
    };

    var temp = args[0].slice(31)
    await stub.putState(temp.substring(0,144), Buffer.from(JSON.stringify(peer)));
    console.info('============= END : Create Peer ===========');
  }


  async transaction(stub, args) {
    console.info('============= START : Transaction ===========');
    if (args.length != 4) {
      throw new Error('Incorrect number of arguments. Expecting 4');
    }

    let signature = args[3]

    const verify = crypto.createVerify('SHA256');
    verify.write(peer1.public_key);
    verify.end();

    if(verify.verify(peer1.public_key,signature)){
      throw new Error('Assinatura não verificada')
    }

    let peerAsBytes1 = await stub.getState(args[0]);
    let peer1 = JSON.parse(peerAsBytes1);
    if(peer1.value < args[2]){
      throw new Error('Balanço insuficiente');
    }
    if(args[2] < 0.0000000001){
      throw new Error('Transferência não suportada: valor demasiado pequeno');
    }
    peer1.value -= args[2];

    let peerAsBytes2 = await stub.getState(args[1]);
    let peer2 = JSON.parse(peerAsBytes2);
    peer2.value += args[2];

    await stub.putState(args[0], Buffer.from(JSON.stringify(peer1)));
    await stub.putState(args[1], Buffer.from(JSON.stringify(peer2)));
    console.info('============= END : Transaction ===========');
  }
};

shim.start(new Chaincode());
